lib/compress.js

```diff
@@ -323,7 +323,7 @@ merge(Compressor.prototype, {
                     if (side_effects_encountered |= lvalues_encountered) continue;
 
                     // Non-constant single use vars can only be replaced in same scope.
-                    if (ref.scope !== self) {
+                    if (ref.scope.get_defun_scope() !== self) {
```

Comment:
Your patch is probably fine if the tests pass, although collapse_vars may be slower as result. But the need for the fix points to a problem with figure_out_scope() in the harmony branch.